File: src/main/java/at/flauschigesalex/defaultLibrary/http/HttpHandler.java
package at.flauschigesalex.defaultLibrary.http;

import lombok.SneakyThrows;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

public abstract class HttpHandler {

    public static HttpResponse<String> get(final @NotNull CharSequence uri) {
        return get(uri, null);
    }

    public static HttpResponse<String> get(final @NotNull CharSequence uri, final @Nullable Map<String, Object> headers) {
        return get(URI.create(uri.toString()), headers);
    }

    public static HttpResponse<String> get(final @NotNull URI uri) {
        return get(uri, null);
    }

    public static HttpResponse<String> get(final @NotNull URI uri, final @Nullable Map<String, Object> headers) {
        return request(uri, HttpRequestType.GET, headers, null);
    }

    public static HttpResponse<String> put(final @NotNull CharSequence uri, final @NotNull Object send) {
        return put(uri, null, send);
    }

    public static HttpResponse<String> put(final @NotNull CharSequence uri, final @Nullable Map<String, Object> headers, final @NotNull Object send) {
        return put(URI.create(uri.toString()), headers, send);
    }

    public static HttpResponse<String> put(final @NotNull URI uri, final @NotNull Object send) {
        return put(uri, null, send);
    }

    public static HttpResponse<String> put(final @NotNull URI uri, final @Nullable Map<String, Object> headers, final @NotNull Object send) {
        return request(uri, HttpRequestType.PUT, headers, send);
    }

    public static HttpResponse<String> post(final @NotNull CharSequence uri, final @NotNull Object send) {
        return post(uri, null, send);
    }

    public static HttpResponse<String> post(final @NotNull CharSequence uri, final @Nullable Map<String, Object> headers, final @NotNull Object send) {
        return post(URI.create(uri.toString()), headers, send);
    }

    public static HttpResponse<String> post(final @NotNull URI uri, final @NotNull Object send) {
        return post(uri, null, send);
    }

    public static HttpResponse<String> post(final @NotNull URI uri, final @Nullable Map<String, Object> headers, final @NotNull Object send) {
        return request(uri, HttpRequestType.POST, headers, send);
    }

    @SneakyThrows
    private static HttpResponse<String> request(final @NotNull URI uri, final @NotNull HttpRequestType type, @Nullable Map<String, Object> headers, final @Nullable Object send) {
        if (headers == null)
            headers = Map.of();

        final HttpRequest.Builder builder = HttpRequest.newBuilder(uri);
        headers.forEach((string, object) -> builder.header(string, object.toString()));
        builder.method(type.name(), send == null ? HttpRequest.BodyPublishers.noBody() : HttpRequest.BodyPublishers.ofString(send.toString()));

        return HttpClient.newHttpClient().send(builder.build(), HttpResponse.BodyHandlers.ofString());
    }
}

File: src/main/java/at/flauschigesalex/defaultLibrary/http/HttpHandler.kt
package at.flauschigesalex.defaultLibrary.http

import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse

@Suppress("unused", "MemberVisibilityCanBePrivate")
class HttpHandler {
    companion object {

        @JvmStatic fun get(url: Any, headers: Map<String, Any> = mapOf()): HttpResponse<String> {
            return get(URI(url.toString()), headers)
        }
        @JvmStatic fun get(uri: URI, headers: Map<String, Any> = mapOf()): HttpResponse<String> {
            return request(uri, HttpRequestType.GET, headers)
        }

        @JvmStatic fun put(uri: Any, headers: Map<String, Any> = mapOf(), send: Any? = null): HttpResponse<String> {
            return put(URI(uri.toString()), headers, send)
        }
        @JvmStatic fun put(uri: URI, headers: Map<String, Any> = mapOf(), send: Any? = null): HttpResponse<String> {
            return request(uri, HttpRequestType.PUT, headers, send)
        }

        @JvmStatic fun post(uri: Any, headers: Map<String, Any> = mapOf(), send: Any? = null): HttpResponse<String> {
            return post(URI(uri.toString()), headers, send)
        }
        @JvmStatic fun post(uri: URI, headers: Map<String, Any> = mapOf(), send: Any? = null): HttpResponse<String> {
            return request(uri, HttpRequestType.POST, headers, send)
        }

        private fun request(
            uri: URI,
            type: HttpRequestType,
            headers: Map<String, Any> = mapOf(),
            send: Any? = null,
        ): HttpResponse<String> {

            val builder = HttpRequest.newBuilder(uri)
            headers.forEach {
                    (string: String, obj: Any),
                -> builder.header(string, obj.toString()) }

            builder.method(
                type.name,
                if (send == null) HttpRequest.BodyPublishers.noBody()
                else HttpRequest.BodyPublishers.ofString(send.toString())
            )

            return HttpClient.newHttpClient().send(builder.build(), HttpResponse.BodyHandlers.ofString())
        }
    }

    enum class HttpRequestType {
        GET, PUT, POST;
    }
}